packages/zoe/src/zoe.js

```diff
@@ -231,38 +299,41 @@ const makeZoe = (additionalEndowments = {}) => {
      * other information, such as the terms used in the instance.
      * @param  {object} installationHandle - the unique handle for the
      * installation
+     * @param  {object} issuerKeywordRecord - a record mapping keyword keys to
+     * issuer values
      * @param  {object} terms - arguments to the contract. These
-     * arguments depend on the contract, apart from the `issuers`
-     * property, which is required.
+     * arguments depend on the contract.
      */
-    makeInstance: (installationHandle, userProvidedTerms) => {
+    makeInstance: (
+      installationHandle,
+      issuerKeywordRecord,
+      terms = harden({}),
+    ) => {
       const { installation } = installationTable.get(installationHandle);
       const instanceHandle = harden({});
       const contractFacet = makeContractFacet(instanceHandle);
 
-      const makeContractInstance = issuerRecords => {
-        const terms = {
-          ...userProvidedTerms,
-          issuers: issuerRecords.map(record => record.issuer),
-        };
+      const keywords = getKeywords(issuerKeywordRecord);
+      const issuers = keywords.map(keyword => issuerKeywordRecord[keyword]);
+      // Take the cleaned keywords and produce a safe(r) issuerKeywordRecord
+      const cleanedRecord = makeCleanedObj(issuerKeywordRecord, keywords);
 
+      const makeInstanceRecord = issuerRecords => {
         const instanceRecord = harden({
           installationHandle,
           publicAPI: undefined,
-          issuers: terms.issuers,
           terms,
+          issuers: issuerRecords.map(record => record.issuer),
+          issuerKeywordRecord: cleanedRecord,
```

Comment:
Are you purposely using `cleanRecord` here, which is stale, rather than the `issuerRecords` parameter, which is called with a non-stale one?

If this is a bug, it is an example of the stale state bug that both @dtribble and I worry about. Here, it occurred using only `then` and explicit promises, rather than `await`, supporting Dean's side ;)

Reply:
What would be stale about `cleanedRecord`?

Reply:
Ah, I see, the `issuers` could be promises for issuers. Let me see what I can do.